File: repack.go
package main

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	"github.com/pkg/errors"
)

type Repack struct {
	Manifest           string
	SourcePath         string
	TargetDirectory    string
	Group              string
	PackageName        string
	Version            string
	Title              string
	PackageDescription string
	IconUrl            string
	Overwrite          bool
}

func (*Repack) Name() string { return "repack" }
func (*Repack) Description() string {
	return "Creates a new ProGet universal package from an existing package with optionally modified metadata."
}

func (r *Repack) Help() string  { return defaultCommandHelp(r) }
func (r *Repack) Usage() string { return defaultCommandUsage(r) }

func (*Repack) PositionalArguments() []PositionalArgument {
	return []PositionalArgument{
		{
			Name:        "source",
			Description: "The path of the existing upack file.",
			Index:       0,
			TrySetValue: trySetPathValue("source", func(cmd Command) *string {
				return &cmd.(*Repack).SourcePath
			}),
		},
	}
}

func (*Repack) ExtraArguments() []ExtraArgument {
	return []ExtraArgument{
		{
			Name:        "manifest",
			Alias:       []string{"metadata"},
			Description: "Path of upack.json file to merge.",
			TrySetValue: trySetPathValue("manifest", func(cmd Command) *string {
				return &cmd.(*Repack).Manifest
			}),
		},
		{
			Name:        "targetDirectory",
			Description: "Directory where the .upack file will be created. If not specified, the current working directory is used.",
			TrySetValue: trySetPathValue("targetDirectory", func(cmd Command) *string {
				return &cmd.(*Repack).TargetDirectory
			}),
		},
		{
			Name:        "group",
			Description: "Package group. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("group", func(cmd Command) *string {
				return &cmd.(*Repack).Group
			}),
		},
		{
			Name:        "name",
			Description: "Package name. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("name", func(cmd Command) *string {
				return &cmd.(*Repack).PackageName
			}),
		},
		{
			Name:        "version",
			Description: "Package version. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("version", func(cmd Command) *string {
				return &cmd.(*Repack).Version
			}),
		},
		{
			Name:        "title",
			Description: "Package title. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("title", func(cmd Command) *string {
				return &cmd.(*Repack).Title
			}),
		},
		{
			Name:        "description",
			Description: "Package description. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("description", func(cmd Command) *string {
				return &cmd.(*Repack).PackageDescription
			}),
		},
		{
			Name:        "icon",
			Description: "Icon absolute Url. If metadata file is provided, value will be ignored.",
			TrySetValue: trySetStringValue("icon", func(cmd Command) *string {
				return &cmd.(*Repack).IconUrl
			}),
		},
		{
			Name:        "overwrite",
			Description: "Overwrite existing package file if it already exists.",
			Flag:        true,
			TrySetValue: trySetBoolValue("overwrite", func(cmd Command) *bool {
				return &cmd.(*Repack).Overwrite
			}),
		},
	}
}

func (r *Repack) Run() int {
	info, err := GetPackageMetadata(r.SourcePath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	infoToMerge, err := r.GetMetadataToMerge()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	prop := func(dest *string, src string) {
		if src != "" {
			*dest = src
		}
	}
	prop(&info.Group, infoToMerge.Group)
	prop(&info.Name, infoToMerge.Name)
	prop(&info.Version, infoToMerge.Version)
	prop(&info.Title, infoToMerge.Title)
	prop(&info.Description, infoToMerge.Description)
	prop(&info.IconURL, infoToMerge.IconURL)
	if len(infoToMerge.Dependencies) != 0 {
		info.Dependencies = infoToMerge.Dependencies
	}
	err = ValidateManifest(info)
	if err != nil {
		thing := "upack.json:"
		if strings.TrimSpace(r.Manifest) == "" {
			thing = "parameters:"
		}
		fmt.Fprintln(os.Stderr, "Invalid", thing, err)
		return 2
	}

	PrintManifest(info)

	relativePackageFileName := info.Name + "-" + info.BareVersion() + ".upack"
	targetFileName, err := filepath.Abs(filepath.Join(r.TargetDirectory, relativePackageFileName))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	if !r.Overwrite {
		_, err = os.Stat(targetFileName)
		if err != nil {
			if os.IsNotExist(err) {
				fmt.Fprintf(os.Stderr, "Target file '%s' exists and overwrite was set to false.", targetFileName)
				return 1
			}
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
	}

	tmpFile, err := ioutil.TempFile("", "upack")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	tmpPath := tmpFile.Name()
	defer func() {
		if tmpFile != nil {
			_ = tmpFile.Close()
			_ = os.Remove(tmpPath)
		}
	}()

	existingPackage, err := zip.OpenReader(r.SourcePath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	builder := zip.NewWriter(tmpFile)
	w, err := builder.Create("upack.json")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	err = json.NewEncoder(w).Encode(info)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	for _, entry := range existingPackage.File {
		if entry.Name == "upack.json" {
			continue
		}

		w, err = builder.CreateHeader(&entry.FileHeader)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}

		if !entry.Mode().IsDir() {
			stream, err := entry.Open()
			if err != nil {
				fmt.Fprintln(os.Stderr, err)
				return 1
			}

			_, err = io.Copy(w, stream)
			if err != nil {
				_ = stream.Close()
				fmt.Fprintln(os.Stderr, err)
				return 1
			}

			err = stream.Close()
			if err != nil {
				fmt.Fprintln(os.Stderr, err)
				return 1
			}
		}
	}

	err = os.MkdirAll(filepath.Dir(targetFileName), 0755)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	err = os.Remove(targetFileName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	err = tmpFile.Close()
	tmpFile = nil
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	err = os.Rename(targetFileName, tmpPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	return 0
}

func (r *Repack) GetMetadataToMerge() (metadata *UniversalPackageMetadata, err error) {
	if strings.TrimSpace(r.Manifest) != "" {
		return &UniversalPackageMetadata{
			Group:       r.Group,
			Name:        r.PackageName,
			Version:     r.Version,
			Title:       r.Title,
			Description: r.PackageDescription,
			IconURL:     r.IconUrl,
		}, nil
	}
	metadataStream, err := os.Open(r.Manifest)
	if err != nil {
		return nil, errors.Wrapf(err, "The manifest file '%s' does not exist or could not be opened.", r.Manifest)
	}
	defer func() {
		if e := metadataStream.Close(); err == nil {
			err = errors.Wrapf(e, "The manifest file '%s' does not exist or could not be opened.", r.Manifest)
		}
	}()

	metadata, err = ReadManifest(metadataStream)
	if err != nil {
		err = errors.Wrapf(err, "The manifest file '%s' does not exist or could not be opened.", r.Manifest)
	}
	return
}

File: upack/Repack.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Inedo.UPack;
using Inedo.UPack.Packaging;
using Newtonsoft.Json.Linq;

namespace Inedo.ProGet.UPack
{
    [DisplayName("repack")]
    [Description("Creates a new ProGet universal package from an existing package with optionally modified metadata.")]
    public sealed class Repack : Command
    {
        [DisplayName("manifest")]
        [AlternateName("metadata")]
        [Description("Path of upack.json file to merge.")]
        [ExtraArgument]
        [ExpandPath]
        public string Manifest { get; set; }

        [DisplayName("source")]
        [Description("The path of the existing upack file.")]
        [PositionalArgument(0)]
        [ExpandPath]
        public string SourcePath { get; set; }

        [DisplayName("targetDirectory")]
        [Description("Directory where the .upack file will be created. If not specified, the current working directory is used.")]
        [ExtraArgument]
        [ExpandPath]
        public string TargetDirectory { get; set; }

        [DisplayName("group")]
        [Description("Package group. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Group { get; set; }

        [DisplayName("name")]
        [Description("Package name. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Name { get; set; }

        [DisplayName("version")]
        [Description("Package version. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Version { get; set; }

        [DisplayName("title")]
        [Description("Package title. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string Title { get; set; }

        [DisplayName("description")]
        [Description("Package description. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string PackageDescription { get; set; }

        [DisplayName("icon")]
        [Description("Icon absolute Url. If metadata file is provided, value will be ignored.")]
        [ExtraArgument]
        public string IconUrl { get; set; }

        [DisplayName("no-audit")]
        [Description("Do not store audit information in the UPack manifest.")]
        [ExtraArgument]
        [DefaultValue(false)]
        public bool NoAudit { get; set; }

        [DisplayName("note")]
        [Description("A description of the purpose for creating this upack file.")]
        [ExtraArgument]
        public string Note { get; set; }

        [DisplayName("overwrite")]
        [Description("Overwrite existing package file if it already exists.")]
        [ExtraArgument]
        [DefaultValue(false)]
        public bool Overwrite { get; set; }

        public override async Task<int> RunAsync(CancellationToken cancellationToken)
        {
            if (this.NoAudit && !string.IsNullOrEmpty(this.Note))
            {
                Console.Error.WriteLine("--no-audit cannot be used with --note.");
                return 2;
            }

            var info = GetPackageMetadata(this.SourcePath);
            var infoToMerge = await GetMetadataToMergeAsync();
            var hash = GetSHA1(this.SourcePath);

            var id = (string.IsNullOrEmpty(info.Group) ? "" : info.Group + "/") + info.Name + ":" + info.Version + ":" + hash;

            foreach (var modifiedProperty in infoToMerge)
                info[modifiedProperty.Key] = modifiedProperty.Value;

            var error = ValidateManifest(info);
            if (error != null)
            {
                Console.Error.WriteLine("Invalid {0}: {1}", string.IsNullOrWhiteSpace(this.Manifest) ? "parameters" : "upack.json", error);
                return 2;
            }

            PrintManifest(info);

            if (!this.NoAudit)
            {
                JArray history;
                if (info.ContainsKey("repackageHistory"))
                {
                    history = (JArray)info["repackageHistory"];
                }
                else
                {
                    history = new JArray();
                    info["repackageHistory"] = history;
                }

                var entry = new Dictionary<string, object>
                {
                    { "id", id },
                    { "date", DateTime.UtcNow.ToString("u") },
                    { "using", "upack/" + typeof(Repack).Assembly.GetName().Version.ToString(3) },
                    { "by", Environment.UserName }
                };

                if (!string.IsNullOrEmpty(this.Note))
                {
                    entry["reason"] = this.Note;
                }

                history.Add(JObject.FromObject(entry));
            }

            string relativePackageFileName = $"{info.Name}-{info.Version.Major}.{info.Version.Minor}.{info.Version.Patch}.upack";
            string targetFileName = Path.Combine(this.TargetDirectory ?? Environment.CurrentDirectory, relativePackageFileName);

            if (!this.Overwrite && File.Exists(targetFileName))
                throw new UpackException($"Target file '{targetFileName}' exists and overwrite was set to false.");

            string tmpPath = Path.GetTempFileName();

            using (var existingPackage = new UniversalPackage(this.SourcePath))
            using (var builder = new UniversalPackageBuilder(tmpPath, info))
            {
                var entries = from e in existingPackage.Entries
                              where !string.Equals(e.RawPath, "upack.json", StringComparison.OrdinalIgnoreCase)
                              select e;

                foreach (var entry in entries)
                {
                    cancellationToken.ThrowIfCancellationRequested();

                    if (entry.IsDirectory)
                    {
                        builder.AddEmptyDirectoryRaw(entry.RawPath);
                    }
                    else
                    {
                        using (var stream = entry.Open())
                        {
                            await builder.AddFileRawAsync(stream, entry.RawPath, entry.Timestamp, cancellationToken);
                        }
                    }
                }
            }

            Directory.CreateDirectory(Path.GetDirectoryName(targetFileName));
            File.Delete(targetFileName);
            File.Move(tmpPath, targetFileName);

            return 0;
        }

        private async Task<UniversalPackageMetadata> GetMetadataToMergeAsync()
        {
            if (string.IsNullOrWhiteSpace(this.Manifest))
            {
                return new UniversalPackageMetadata
                {
                    Group = this.Group,
                    Name = this.Name,
                    Version = UniversalPackageVersion.TryParse(this.Version),
                    Title = this.Title,
                    Description = this.PackageDescription,
                    Icon = this.IconUrl
                };
            }
            else
            {
                try
                {
                    using (var metadataStream = File.OpenRead(this.Manifest))
                    {
                        return await ReadManifestAsync(metadataStream);
                    }
                }
                catch (Exception ex)
                {
                    throw new UpackException($"The manifest file '{this.Manifest}' does not exist or could not be opened.", ex);
                }
            }
        }
    }
}
